### PR TITLE
fix: ensure Java interopt

### LINKED ISSUE #98: Test against java kafka client
We currently have a test matrix that produces and consumes data using `rdkafka` and `rskafka` in all supported compression formats:

https://github.com/influxdata/rskafka/blob/0f9cd4ebe802d450410b0307ff9a85d814147a17/tests/produce_consume.rs#L16-L106

Since record batches are mostly only processed by the clients, it would be nice to add an official java/scala client to test matrix via some bridge / CLI call, just to make sure we can work with most of the ecosystem.

### PR DESCRIPTION
**Pull request by crepererum**
The only issue found by this is that we were not able to read
Snappy-compressed data produced by Java clients, because Java uses its
own weird framing format. Funnily Java clients can read the Snappy data
what we produce. This is also in line with the rdkafka implemention
(they can consume the "normal" and the special "Java" format but only
produce the "normal" one).

Closes #98.

### CODE REVIEW

File: tests/java_helper.rs [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+            )
+            .expect("creating KafkaProducer");
+
+        let fut = jvm
+            .invoke(&producer, "send", &[InvocationArg::from(jvm_record)])
+            .expect("flush");
+        futures.push(fut);
+    }
+
+    jvm.invoke(&producer, "flush", &[]).expect("flush");
```
Comment (tustvold) [inline tests/java_helper.rs:112]:
```suggestion
    // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#flush()
    jvm.invoke(&producer, "flush", &[]).expect("flush");
```

File: tests/java_helper.rs [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+            )
+            .expect("cast to RecordMetadata");
+        let offset = jvm
+            .invoke(&record_metadata, "offset", &[])
+            .expect("getting offset");
+        let offset: i64 = jvm.to_rust(offset).expect("converting offset to Rust");
+        offsets.push(offset);
+    }
+
+    jvm.invoke(&producer, "close", &[]).expect("close");
```
Comment (tustvold) [inline tests/java_helper.rs:130]:
```suggestion
    // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#close()
    jvm.invoke(&producer, "close", &[]).expect("close");
```

File: tests/java_helper.rs [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+            ),
+            (
+                "value.deserializer",
+                "org.apache.kafka.common.serialization.StringDeserializer",
+            ),
+            ("auto.offset.reset", "earliest"),
+        ],
+    );
+
+    let consumer = jvm
```
Comment (tustvold) [inline tests/java_helper.rs:160]:
```suggestion
    // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#%3Cinit%3E(java.util.Map)
    let consumer = jvm
```

File: tests/java_helper.rs [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    connection: &str,
+    topic_name: &str,
+    partition_index: i32,
+    n: usize,
+) -> Vec<RecordAndOffset> {
+    let jvm = setup_jvm();
+
+    let props = create_properties(
+        &jvm,
+        &[
```
Comment (tustvold) [inline tests/java_helper.rs:146]: Might want to set `enable.auto.commit` to `false`

File: tests/java_helper.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        )
+        .expect("creating partitions array");
+    let partitions = jvm
+        .invoke_static(
+            "java.util.Arrays",
+            "asList",
+            &[InvocationArg::from(partitions)],
+        )
+        .expect("partitions asList");
+    jvm.invoke(&consumer, "assign", &[InvocationArg::from(partitions)])
```
Comment (tustvold) [inline tests/java_helper.rs:192]:
```suggestion
    // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#assign(java.util.Collection)
    jvm.invoke(&consumer, "assign", &[InvocationArg::from(partitions)])
```

File: tests/java_helper.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        .expect("assign");
+
+    let mut results = vec![];
+    while results.len() < n {
+        let consumer_records = jvm
```
Comment (tustvold) [inline tests/java_helper.rs:197]:
```suggestion
        // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#poll(long)
        let consumer_records = jvm
```

File: tests/java_helper.rs [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+                &consumer,
+                "poll",
+                &[InvocationArg::try_from(1_000i64)
+                    .expect("timeout arg")
+                    .into_primitive()
+                    .expect("timeout into primitive")],
+            )
+            .expect("poll");
+
+        let it = jvm
```
Comment (tustvold) [inline tests/java_helper.rs:208]:
```suggestion
        // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/ConsumerRecords.html#iterator()
        let it = jvm
```

File: tests/java_helper.rs [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+            .invoke(&consumer_records, "iterator", &[])
+            .expect("iterator");
+        for consumer_record in JavaIterator::new(&jvm, it) {
+            let consumer_record = jvm
```
Comment (tustvold) [inline tests/java_helper.rs:212]: Not sure I understand this cast either

File: tests/java_helper.rs [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+                .cast(
+                    &consumer_record,
+                    "org.apache.kafka.clients.consumer.ConsumerRecord",
+                )
+                .expect("cast to ConsumerRecord");
+
+            let key = jvm.invoke(&consumer_record, "key", &[]).expect("key");
```
Comment (tustvold) [inline tests/java_helper.rs:219]:
```suggestion
            // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/ConsumerRecord.html
            let key = jvm.invoke(&consumer_record, "key", &[]).expect("key");
```

File: tests/java_helper.rs [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                headers,
+                timestamp: OffsetDateTime::from_unix_timestamp_nanos(timestamp as i128 * 1_000_000)
+                    .unwrap(),
+            };
+            let record_and_offset = RecordAndOffset { record, offset };
+            results.push(record_and_offset);
+        }
+    }
+
+    jvm.invoke(&consumer, "close", &[]).expect("close");
```
Comment (tustvold) [inline tests/java_helper.rs:272]:
```suggestion
    // https://kafka.apache.org/31/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#close()
    jvm.invoke(&consumer, "close", &[]).expect("close");
```

File: tests/java_helper.rs [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    fn new(jvm: &'a Jvm, it: Instance) -> Self {
+        Self { jvm, it }
+    }
+}
+
+impl<'a> Iterator for JavaIterator<'a> {
+    type Item = Instance;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let has_next = self.jvm.invoke(&self.it, "hasNext", &[]).expect("hasNext");
```
Comment (tustvold) [inline tests/java_helper.rs:391]:
```suggestion
        // https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Iterator.html#hasNext()
        let has_next = self.jvm.invoke(&self.it, "hasNext", &[]).expect("hasNext");
```

File: tests/java_helper.rs [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+        let has_next: bool = self.jvm.to_rust(has_next).expect("hasNext to Rust");
+        if has_next {
+            Some(self.jvm.invoke(&self.it, "next", &[]).expect("next"))
```
Comment (tustvold) [inline tests/java_helper.rs:394]:
```suggestion
            // https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Iterator.html#next()
            Some(self.jvm.invoke(&self.it, "next", &[]).expect("next"))
```

**Review by tustvold — APPROVED** [review]
Very cool :+1:

File: src/protocol/record.rs [hunk 13]
```diff
@@ -619,49 +619,39 @@ where
...
+                // There are "normal" compression libs, and there is Java
+                // See https://github.com/edenhill/librdkafka/blob/2b76b65212e5efda213961d5f84e565038036270/src/rdkafka_msgset_reader.c#L307-L318
+                let output = if input.starts_with(JAVA_MAGIC) {
+                    let mut cursor = Cursor::new(&input[JAVA_MAGIC.len()..]);
+
+                    let mut buf_version = [0u8; 4];
+                    cursor.read_exact(&mut buf_version)?;
+
+                    let mut buf_compatible = [0u8; 4];
+                    cursor.read_exact(&mut buf_compatible)?;
```
[the remark this answers is not in the document]
Comment (crepererum) [reply src/protocol/record.rs]: done

File: tests/java_helper.rs (hunk 8 above)
Comment (crepererum) [reply tests/java_helper.rs:228]: added some docs

File: tests/java_helper.rs [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+            .expect("flush");
+        futures.push(fut);
+    }
+
+    jvm.invoke(&producer, "flush", &[]).expect("flush");
+
+    let mut offsets = vec![];
+    for fut in futures {
+        let record_metadata = jvm.invoke(&fut, "get", &[]).expect("polling future");
+        let record_metadata = jvm
```
[the remark this answers is not in the document]
Comment (crepererum) [reply tests/java_helper.rs:127]: added some docs

File: tests/java_helper.rs (hunk 4 above)
Comment (crepererum) [reply tests/java_helper.rs:160]: done

File: tests/java_helper.rs [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+            ],
+        )
+        .expect("creating TopicPartition");
+    let partitions = jvm
+        .create_java_array(
+            "org.apache.kafka.common.TopicPartition",
+            &[InvocationArg::from(topic_partition)],
+        )
+        .expect("creating partitions array");
+    let partitions = jvm
```
[the remark this answers is not in the document]
Comment (crepererum) [reply tests/java_helper.rs]: good idea! done.

**Pull request merged**